### PR TITLE
pgwire: add server.max_connections public cluster setting

### LINKED ISSUE #54653: SQL: add max open connections
Similarly to https://github.com/cockroachdb/cockroach/issues/54652, we should add a cluster setting and session variable that allows users to set a MAX OPEN CONNECTIONS. This can help provide rough protection to the database from excessive opened connections. The value would default to infinite and be configurable to users. If a connection would be opened against the database that exceeds the number of max connections, we would return a unique error to the clients.

### PR DESCRIPTION
**Pull request by asubiotto**
This setting is disabled by default. If non-zero, this setting specifies a
maximum number of connections that a server can have open at any given time.
If a new connection would exceed this limit, the same error message is
returned as postgres: "sorry, too many connections" with the 53300 error code
that corresponds to "too many connections".

Release note (sql change): an off-by-default server.max_connections cluster
setting has been added to limit the maximum number of connections to a server.

Closes #54653 

cc @awoods187 

Example error from the SQL shell:
```
./cockroach sql --insecure
Flag --insecure has been deprecated, it will be removed in a subsequent release.
For details, see: https://go.crdb.dev/issue-v/53404/v20.2
#
# Welcome to the CockroachDB SQL shell.
# All statements must be terminated by a semicolon.
# To exit, type: \q.
#
ERROR: sorry, too many clients already
SQLSTATE: 53300
HINT: the maximum number of allowed connections is 1 and can be modified using the server.max_connections cluster setting
Failed running "sql"
```

### CODE REVIEW

**Comment by awoods187** [comment]
I believe Postgres usually sets this as a default to -1 which is infinite as opposed to not turning it on. Does this error message match Postgres (its hard for me to tell)?

**Comment by asubiotto** [comment]
Postgres sets this to 115 as a default with 15 reserved for the superuser. Yes, the error message matches postgres: https://stackoverflow.com/questions/2757549/org-postgresql-util-psqlexception-fatal-sorry-too-many-clients-already
 
>  which is infinite as opposed to not turning it on

The default here is 0, which means infinite. What do you mean by not turning it on vs infinite?

**Comment by knz** [comment]
Can you check the overlap with #51505

including:
- checking whether your PR also addresses #35428
- addressing @bdarnell's remaining concern about the max limit blocking legitimate users out

**Comment by knz** [comment]
Also given this is a cluster setting, what happens for the initial connections, when you have a large number of SQL apps doing a TCP connect on the SQL listener *before the cluster is initialized* ?

**Comment by awoods187** [comment]
[Databases](https://www.postgresql.org/docs/9.1/sql-alterdatabase.html) `-1 means no limit`  and 0 would be none. The default is 115 with 15 reserved for superuser and [100 generally available](https://www.postgresql.org/docs/9.4/runtime-config-connection.html).
[Roles](https://www.postgresql.org/docs/9.1/sql-alterdatabase.html) offers the same -1 and 0 semantics however its default is -1 and can be set to any number.

**Review by andreimatei — COMMENTED** [review]
The limit is per gateway, right? I think the name should express that somehow, but I'm not sure how. `server.max_connections_per_gateway`?

What's the story about increasing this limit once you've hit it? Do we also reserve some connections for some users like pg? The commit msg doesn't say.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54745#-:-MI-RQTWEd9b-R8Id9Ph:b-k1sppu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @jordanlewis, and @knz)

**Comment by asubiotto** [comment]
FYI this is something we'd like to backport to 20.2 because it's an explicit user request.
cc @a-entin 

> checking whether your PR also addresses #35428

It does.

> addressing @bdarnell's remaining concern about the max limit blocking legitimate users out

It doesn't address it because this is done post-TLS establishment but pre-auth. I can investigate doing this post-auth and simply not have a limit for the root user. I wonder what @awoods187 thinks.

> Also given this is a cluster setting, what happens for the initial connections, when you have a large number of SQL apps doing a TCP connect on the SQL listener before the cluster is initialized ?

The flood of new connections will be accepted until the setting is gossiped. Happy to change this to a flag.

@awoods187, the -1 semantics do not apply to `max_connections`:
```
asubiotto=# ALTER system set max_connections=-1;
ERROR:  -1 is outside the valid range for parameter "max_connections" (1 .. 262143)
```

**Comment by andreimatei** [comment]
Reading the comments in #51505 I agree with what the thinking there seems to be (or at least how I read it) - the rate of opening conns is a bigger problem than the absolute number of open connections, so I think that if we only get one limit it should be that one.
One way or another I think we need to allow an administrator to connect even when the limits are hit. Does this patch apply to connections opened over the unix domain socket? I saw that the one attempted in #51505 did not - and I think that was a good idea? So if we make domain socket conns exempt from all the limits, that'd be part of the answer. For TCP conns (and TLS conns in particular), I'm not sure what can be done pre-auth. If we have the username pre-auth, then we can have limits per user (even if that allows a DoS using admin usernames to starve real admin conns). If we don't, then I guess the absolute limit can be enforced post-auth, and I'm not sure what we do with the rate limit.

**Comment by asubiotto** [comment]
This PR does not attempt to solve a hotspot problem. What it aims to do is allow operators to set a max limit of connections they are comfortable having their system handle. i.e. we're not talking about rogue apps, but mostly well-behaved ones that are disconnected from system limitations cc @a-entin for clarification if I'm getting things wrong

Agreed with the administrator, this was an oversight on my part. This patch does apply to connections on the unix domain socket but I think I'll exempt admin users.

To summarize I'm going to change this PR in two ways:
1) Exempt admin users from these limits. I'll do this post-auth since this PR's aim is to be correct, not performant.
2) Change the cluster setting to a flag so that uninitialized servers can still enforce a limit.

**Comment by bdarnell** [comment]
> the rate of opening conns is a bigger problem than the absolute number of open connections, so I think that if we only get one limit it should be that one.

When we have SCRAM auth, I'll be a lot less concerned about the rate limit. But they're really separate problems, and you need both to guard against DoS. I think the connection limit is probably the more important one right now because without other safeguards (that we don't have yet), it's relatively easy for anyone with network access to run the process into its file descriptor limit and probably crash it. 

> Does this patch apply to connections opened over the unix domain socket? I saw that the one attempted in #51505 did not - and I think that was a good idea? So if we make domain socket conns exempt from all the limits, that'd be part of the answer.

That's one possible method for emergency access, but unix domain sockets are really hard to use from many deployments (e.g. k8s). Another option would be to listen on multiple SQL ports with separate connection limits, so that firewall rules could limit access to the emergency port (see #54006 and #51453). 

Consider making this a command-line flag instead of a cluster setting - this would be another way to undo a problematic configuration in an emergency. 

I'm 👎  on backporting this - we really need to take a release to think more holistically about the growing ad-hoc collection of configuration options here.

**Comment by asubiotto** [comment]
Made the cluster setting a flag and moved the check post-auth to allow admins to bypass this limit.

**Comment by a-entin** [comment]
Whatever the implementation will end up being (config vs start up flag based) an important aspect is that the cluster wide setting needs to be adjustable without requiring a restart of any node.

The scenario to keep in mind - in response to workload increase IT adds nodes and needs to increase the max connections accordingly. Similar re downsizing. Can think of other plausible operating scenarios.

**Comment by asubiotto** [comment]
Closing as stale
